src/core/algorithms/cfd/cfdfinder/model/cfdfinder_relation_data.cpp

```diff
@@ -0,0 +1,86 @@
+#include "cfdfinder_relation_data.h"
+
+#include <deque>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <easylogging++.h>
+
+namespace {
+std::unique_ptr<model::PositionListIndex> FetchPLI(algos::cfdfinder::ClusterMap& cluster_map,
+                                                   bool is_null_equal_null) {
+    std::deque<std::vector<int>> clusters;
+    std::vector<int> null_cluster;
+    unsigned int size = 0;
+    unsigned int relation_size = 0;
+    static std::string const kNullValue = "";
+
+    if (cluster_map.contains(kNullValue)) {
+        null_cluster = cluster_map[kNullValue];
+    }
+    if (!is_null_equal_null) {
+        cluster_map.erase(kNullValue);
+        relation_size += null_cluster.size();
+    }
+    for (auto& iter : cluster_map) {
+        size += iter.second.size();
+        if (iter.second.size() == 1) {
+            continue;
+        }
+        clusters.push_back(iter.second);
+    }
+    relation_size += size;
+    std::sort(clusters.begin(), clusters.end(),
+              [](std::vector<int> const& a, std::vector<int> const& b) { return a[0] < b[0]; });
```

Comment:
FYI: `ranges` can be used for brevity:
```suggestion
    std::ranges::sort(clusters, {}, [](std::vector<int> const& a) { return a[0]; });
```